internal/controlplane/handlers_repositories.go

```diff
@@ -46,106 +45,40 @@ const maxFetchLimit = 100
 // Once a user had enrolled in a project (they have a valid token), they can register
 // repositories to be monitored by the minder by provisioning a webhook on the
 // repository(ies).
-func (s *Server) RegisterRepository(ctx context.Context,
-	in *pb.RegisterRepositoryRequest) (*pb.RegisterRepositoryResponse, error) {
-	entityCtx := engine.EntityFromContext(ctx)
-	projectID := entityCtx.Project.ID
-
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, projectID)
-	if err != nil {
-		return nil, providerError(err)
-	}
-
-	pbOpts := []providers.ProviderBuilderOption{
-		providers.WithProviderMetrics(s.provMt),
-		providers.WithRestClientCache(s.restClientCache),
-	}
-	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, pbOpts...)
+func (s *Server) RegisterRepository(
+	ctx context.Context,
+	in *pb.RegisterRepositoryRequest,
+) (*pb.RegisterRepositoryResponse, error) {
+	projectID := getProjectID(ctx)
+	provider, client, err := s.getProviderAndClient(ctx, projectID, in)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+		return nil, err
 	}
 
-	// Unmarshal the in.GetRepositories() into a struct Repository
-	if in.GetRepository() == nil || in.GetRepository().Name == "" {
-		return nil, util.UserVisibleError(codes.InvalidArgument, "no repository provided")
+	// Validate that the Repository struct in the request
+	repoReference := in.GetRepository()
+	if repoReference == nil || repoReference.Name == "" || repoReference.Owner == "" {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "missing repository owner and/or name")
 	}
 
-	repo := in.GetRepository()
-
-	result, err := s.registerWebhookForRepository(ctx, p, projectID, repo)
+	newRepo, err := s.repos.CreateRepository(ctx, client, provider, projectID, repoReference)
 	if err != nil {
-		return nil, util.UserVisibleError(codes.Internal, "cannot register webhook: %v", err)
+		if errors.Is(err, ghrepo.ErrPrivateRepoForbidden) {
```

Comment:
Generally I prefer to avoid API-specific details from the business logic, but I may consider changing that approach in future in order to avoid creating lots of custom error types for the API to switch on.